src/conman/core.clj

```diff
@@ -64,14 +71,14 @@
        (doseq [[id# {fn# :fn {doc# :doc} :meta}] fns#]
          (intern *ns* (with-meta (symbol (name id#)) {:doc doc#})
                  (fn f#
-                   ([] (f# (if (fn? ~conn) (~conn) ~conn) {}))
-                   ([params#] (f# (if (fn? ~conn) (~conn) ~conn) params#))
+                   ([] (f# (resolve-conn ~conn) {}))
```

Comment:
It would be better to call `resolve-conn` in the let binding so that it happens once instead of incurring runtime cost on each call.

Reply:
Do you mean something like this?

```clj
(defmacro bind-connection [conn & filenames]
  (let [options?  (map? (first filenames))
        options   (if options? (first filenames) {})
        filenames (if options? (rest filenames) filenames)]
    `(let [{snips# :snips fns# :fns :as queries#} (conman.core/load-queries '~filenames ~options)
           implicit-conn# (resolve-conn ~conn)]
       (doseq [[id# {fn# :fn {doc# :doc} :meta}] snips#]
         (intern *ns* (with-meta (symbol (name id#)) {:doc doc#})
                 (fn [& args#]
                   (try (apply fn# args#)
                        (catch Exception e#
                          (throw (Exception. (format "Exception in %s" id#) e#)))))))
       (doseq [[id# {fn# :fn {doc# :doc} :meta}] fns#]
         (intern *ns* (with-meta (symbol (name id#)) {:doc doc#})
                 (fn f#
                   ([] (f# implicit-conn# {}))
                   ([params#] (f# implicit-conn# params#))
                   ([conn# params#]
                    (try (fn# (resolve-conn conn#) params#)
                         (catch Exception e#
                           (throw (Exception. (format "Exception in %s" id#) e#)))))
                   ([conn# params# opts# & command-opts#]
                    (try (apply fn# (resolve-conn conn#) params# opts# command-opts#)
                         (catch Exception e#
                           (throw (Exception. (format "Exception in %s" id#) e#))))))))
       queries#)))
```
Tests fail on this one:

```
lein test :only conman.core-test/transaction

FAIL in (transaction) (core_test.clj:81)
expected: (= [] (get-fruit {:name "apple"}))
  actual: (not (= [] ({:id 0, :name "apple", :appearance "red", :cost 1, :grade 1})))

lein test conman.core2-test

lein test :only conman.core2-test/transaction

FAIL in (transaction) (core2_test.clj:51)
expected: (= [] (get-fruit {:name "apple"}))
  actual: (not (= [] ({:id 0, :name "apple", :appearance "red", :cost 1, :grade 1})))
```

The let up there is executed at the time of binding, but the resolution has to be made every time `f#` is called. As a result, the transaction `binding` is not used.

Reply:
Or something like this, without `resolve-conf` at all:

```clj
(defmacro bind-connection [conn & filenames]
  (let [options?  (map? (first filenames))
        options   (if options? (first filenames) {})
        filenames (if options? (rest filenames) filenames)]
    `(let [{snips# :snips fns# :fns :as queries#} (conman.core/load-queries '~filenames ~options)]
       (doseq [[id# {fn# :fn {doc# :doc} :meta}] snips#]
         (intern *ns* (with-meta (symbol (name id#)) {:doc doc#})
                 (fn [& args#]
                   (try (apply fn# args#)
                        (catch Exception e#
                          (throw (Exception. (format "Exception in %s" id#) e#)))))))
       (doseq [[id# {fn# :fn {doc# :doc} :meta}] fns#]
         (intern *ns* (with-meta (symbol (name id#)) {:doc doc#})
                 (if (instance? IDeref ~conn)
                   (fn f#
                     ([] (f# (deref ~conn) {}))
                     ([params#] (f# (deref ~conn) params#))
                     ([conn# params#]
                      (try (fn# conn# params#)
                           (catch Exception e#
                             (throw (Exception. (format "Exception in %s" id#) e#)))))
                     ([conn# params# opts# & command-opts#]
                      (try (apply fn# conn# params# opts# command-opts#)
                           (catch Exception e#
                             (throw (Exception. (format "Exception in %s" id#) e#))))))
                   (fn f#
                     ([] (f# ~conn {}))
                     ([params#] (f# ~conn params#))
                     ([conn# params#]
                      (try (fn# conn# params#)
                           (catch Exception e#
                             (throw (Exception. (format "Exception in %s" id#) e#)))))
                     ([conn# params# opts# & command-opts#]
                      (try (apply fn# conn# params# opts# command-opts#)
                           (catch Exception e#
                             (throw (Exception. (format "Exception in %s" id#) e#)))))))
                 ))
       queries#)))
```